frame/contracts/src/exec.rs

```diff
@@ -645,17 +647,25 @@ where
 			<ContractInfoOf<T>>::insert(frame.account_id.clone(), contract.clone());
 		}
 
-		let frame = top_frame_mut!(self);
-		let nested_gas = &mut frame.nested_gas;
-		let nested_storage = &mut frame.nested_storage;
-		let (frame, executable, _) = Self::new_frame(
+		let parent_frame = top_frame_mut!(self);
+		let nested_gas = &mut parent_frame.nested_gas;
+		let nested_storage = &mut parent_frame.nested_storage;
+		let (mut frame, executable, _) = Self::new_frame(
 			frame_args,
 			value_transferred,
 			nested_gas,
 			nested_storage,
 			gas_limit,
 			self.schedule,
 		)?;
+
+		// In order to execute contract in a parent context we need to pass caller account_id
+		// and contract_info to the new frame
+		if preserve_context {
+			frame.account_id = parent_frame.account_id.clone();
```

Comment:
The context should be full. You also need: `value_transferred`, `balance`, `caller` and etc

Reply:
`value_transferred` is passed untouched, `balance` (`storage_deposit`) is passed in contract info, `caller` is always taken from the 1st frame